plugins/handler/collectd-metrics/main.go

```diff
@@ -0,0 +1,153 @@
+package main
+
+import (
+	"time"
+
+	"github.com/go-openapi/errors"
+	"github.com/infrawatch/sg-core-refactor/pkg/data"
+	"github.com/infrawatch/sg-core-refactor/pkg/handler"
+	"github.com/infrawatch/sg-core-refactor/plugins/handler/collectd-metrics/pkg/collectd"
+)
+
+type collectdMetricsHandler struct {
+	totalMetricsReceived uint64 //total number of internal metrics created from collectd blobs
+	totalDecodeErrors    uint64
+}
+
+func (c *collectdMetricsHandler) Handle(blob []byte) []data.Metric {
+
+	var err error
+	var cdmetrics *[]collectd.Metric
+
+	cdmetrics, err = collectd.ParseInputByte(blob)
+	metrics := []data.Metric{}
+
+	if err != nil {
+		c.totalDecodeErrors++
+		return metrics
+	}
+
+	var ms []data.Metric
+	if cdmetrics == nil {
+		c.totalDecodeErrors++
+		return metrics
+	}
+
+	for _, cdmetric := range *cdmetrics {
+		ms, err = c.createMetrics(&cdmetric)
+		if err != nil {
+			c.totalDecodeErrors++
+		}
+		metrics = append(metrics, ms...)
+	}
+
+	metrics = append(metrics, []data.Metric{{
```

Comment:
Yeah that is what I was referring to.

Reply:
Hmmm...  Are inline metrics generated here because the handlers are now separate executables from the Applications?  While this traffic is internal, it could still greatly multiple the number of messages, encoding/decoding.  Depending on the bundling ratio from collectd, the number of sg_total_metric_rcv_count messages could equal the number of metrics sent per period.  It would be acceptable to only include these counters once a second in the stream. (i.e. a second optional append)

Reply:
As we discuss this more, I am beginning to think the internal bus should consist of a large cache of metric objects only created once and then updated by the handlers. It would save the overhead of object allocation and provide effectively the same functionality except that creating the metric would be a function call rather than a object creation. Expiration of metrics could be handled by the bus then. The events may need to stay the same since they are more variable